### PR TITLE
Convert rstan unit testing to use testthat

### PR DESCRIPTION
**Pull request by jeffreypullin**
#### Summary:

* Converts rstan testing infrastructure to use testthat 

* Centralizes all rstan's R tests 

Partially completes #907 

#### Intended Effect:

Using testthat makes it easier for new contributors to add tests helping to increase rstan's test coverage. Coverage has been a problem while trying to land #887, for example. 

While doing the conversion I also found some odd features of the previous testing setup. There were two folders of R tests using {RUnit}: one in `/rstan/rstan/rstan/inst/unitTests` and one in `/rstan/rstan/tests/unitTests/`. The one in `/rstan/rstan/tests/unitTests/` was bigger and the tests 'heavier' i.e. involving fitting small Stan models. Only the tests in `/rstan/rstan/rstan/inst/unitTests` were being routinely run during testing, however, due to a bug/feature in the `runRunitTests.R` file. I have included *all* the tests in the new `testthat` test folder, but those not run routinely in the past are all skipped (at least for now). Gaining a better understanding of the available tests and putting them all in one place has been a productive byproduct of the conversion to testthat. 

I have left the existing cpp tests as well as some files in `inst/unitTests/` that were not {RUnit} test files but were seemingly used for testing. 

#### How to Verify:

* Check that CI works and passes
* Manually verify that tests are equivalent

#### Side Effects:

Reorganization of test files

#### Documentation:

NA

#### Reviewer Suggestions: 

@bgoodri  @hsbadr @rok-cesnovar 

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):

Jeffrey Pullin

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: GPLv3 (http://opensource.org/licenses/GPL-3.0)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

**Comment by rok-cesnovar** [comment]
Thanks @jeffreypullin ! Will get to this in the following days. Sorry for the delayed response.

**Review by rok-cesnovar — APPROVED** [review]
This is fantastic stuff. Thanks @jeffreypullin. @bgoodri do you want to hold of on merging this while you work on 2.26?

**Comment by bgoodri** [comment]
Yeah, we'll merge it into develop. Thanks for looking at it. I'm still
trying to catch up on everything I fell behind on this past week.

**Comment by jeffreypullin** [comment]
Great! Thanks for the review @rok-cesnovar!

**Comment by jeffreypullin** [comment]
Bump!

**Comment by hsbadr** [comment]
@jeffreypullin Could you target the `experimental` branch?

**Comment by jeffreypullin** [comment]
@hsbadr Rebased onto experimental!

File: rstan/rstan/DESCRIPTION [hunk 1]
```diff
@@ -53,7 +53,8 @@ LinkingTo:
     StanHeaders (>= 2.27.0),
     RcppParallel (>= 5.1.4)
 Suggests:
-    RUnit,
+    RcppEigen (>= 0.3.3.3.0),
+    BH (>= 1.72.0-2),
```
Comment (hsbadr) [inline rstan/rstan/DESCRIPTION]: We're already linking to `RcppEigen` and `BH`.
```suggestion
```

File: rstan/rstan/DESCRIPTION [hunk 2]
```diff
@@ -62,10 +63,12 @@ Suggests:
     rstantools,
     rstudioapi,
     Matrix,
-    knitr
-URL: https://mc-stan.org/rstan/, https://discourse.mc-stan.org
+    knitr,
+    V8,
+    testthat (>= 3.0.0)
+URL: https://mc-stan.org/rstan, https://discourse.mc-stan.org
```
Comment (hsbadr) [inline rstan/rstan/DESCRIPTION]: We already import `V8`. So, please just replace `RUnit` with `testthat`. All other changes aren't needed.

File: rstan/rstan/DESCRIPTION [hunk 3]
```diff
@@ -62,10 +63,12 @@ Suggests:
...
 BugReports: https://github.com/stan-dev/rstan/issues/
 VignetteBuilder: knitr
 SystemRequirements: GNU make, pandoc
-Encoding: UTF-8
-RoxygenNote: 7.1.1
+RoxygenNote: 5.0.1
+Config/testthat/edition: 3
```
Comment (hsbadr) [inline rstan/rstan/DESCRIPTION]: Just add `testthat` edition config and reset the other changes.
```suggestion
Encoding: UTF-8
RoxygenNote: 7.1.1
Config/testthat/edition: 3
```

**Comment by hsbadr** [comment]
> @hsbadr Rebased onto experimental!

Thanks @jeffreypullin! LGTM. Just requested minor changes in the `DESCRIPTION` to only replace `RUnit` with `testthat`.

**Comment by jeffreypullin** [comment]
Whoops - sorry about those: I messed up the rebase onto experimental. Should be fixed now. 

Thanks!

**Comment by hsbadr** [comment]
> Whoops - sorry about those: I messed up the rebase onto experimental. Should be fixed now.

Thanks! There's a missing comma after `testthat` in Suggests, and there's no need to change the URL (I don't remember if CRAN checks suggested the training slash).

File: rstan/rstan/DESCRIPTION [hunk 4]
```diff
@@ -53,7 +53,7 @@ LinkingTo:
     StanHeaders (>= 2.27.0),
     RcppParallel (>= 5.1.4)
 Suggests:
-    RUnit,
+    testthat (>= 3.0.0)
```
Comment (hsbadr) [inline rstan/rstan/DESCRIPTION]:
```suggestion
    testthat (>= 3.0.0),
```

File: rstan/rstan/DESCRIPTION [hunk 5]
```diff
@@ -63,9 +63,10 @@ Suggests:
     rstudioapi,
     Matrix,
     knitr
-URL: https://mc-stan.org/rstan/, https://discourse.mc-stan.org
+URL: https://mc-stan.org/rstan, https://discourse.mc-stan.org
```
Comment (hsbadr) [inline rstan/rstan/DESCRIPTION]:
```suggestion
URL: https://mc-stan.org/rstan/, https://discourse.mc-stan.org
```

**Comment by jeffreypullin** [comment]
Done, thanks for the detailed review!

**Review by hsbadr — APPROVED** [review]

**Pull request merged**